File: 16_tree/01-iterative-pre-order-traversal/01-iterative-pre-order-traversal.rb
require_relative '../../RubyDataStructure/BinaryTree/TreeNode.rb'

def iterative_pre_order_traversal(root)
  st, res = [], []
  while !st.empty? || !root.nil?
    if !root.nil?
      st.push(root)
      res.push(root.val)
      root = root.left
    else
      root = st.pop
      root = root.right
    end
  end
  return res
end

root = TreeNode.new(10, TreeNode.new(20), TreeNode.new(30, TreeNode.new(40)))
res = iterative_pre_order_traversal(root)
print res

File: 16_tree/01-iterative-pre-order-traversal/01-iterative-pre-order-traversal.ts
import { TreeNode } from '../../TypeScriptDataStructure/BinaryTree/TreeNode'

const iterative_pre_traversal = (root: TreeNode) => {
  const st = []
  const res = []
  while (st.length > 0 || root != null) {
    if (root != null) {
      st.push(root)
      res.push(root.val)
      root = root.left
    }
    else {
      root = st.pop()
      root = root.right
    }
  }
  return res
}

const root = new TreeNode(10, new TreeNode(20,null, new TreeNode(40)), new TreeNode(30))
console.log(iterative_pre_traversal(root))